src/main/java/org/elasticsearch/cluster/Diffable.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster;
+
+import org.elasticsearch.common.io.stream.ImmutableStreamable;
+import org.elasticsearch.common.io.stream.StreamInput;
+
+import java.io.IOException;
+
+/**
+ * Cluster state part, changes in which can be serialized
+ */
+public interface Diffable<T> extends ImmutableStreamable<T> {
+
+    /**
+     * Returns serializable object representing differences between this and previousState
+     */
+    Diff diffs(T previousState);
+
+    /**
+     * Reads the {@link org.elasticsearch.cluster.Diff} from StreamInput and applies them to this
+     */
+    T readDiffs(StreamInput in) throws IOException;
```

Comment:
maybe call this readDiffAndApply? this doesn't really read a diff and return it.